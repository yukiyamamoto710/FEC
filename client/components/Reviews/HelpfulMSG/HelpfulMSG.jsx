import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import putHelpfulReview from '../func/putHelpfulReview/putHelpfulReview';
import putReportReview from '../func/putReportReview/putReportReview';

const HelpfulMSG = (props) => {
  const {
    countHelpful,
    countNotHelpful,
    reported,
    reviewId,
  } = props;

  const [isClicked, setIsClicked] = useState(false);
  const [countYes, setCountYes] = useState(countHelpful);
  const [countNo, setCountNo] = useState(countNotHelpful);

  useEffect(() => {
    setCountYes(countHelpful);
    setCountNo(countNotHelpful);
    setIsClicked(false);
  }, [reviewId]);

  useEffect(() => {
    if (isClicked) {
      putHelpfulReview({ review_id: Number(reviewId) }, reviewId);
      // axios.put('/reviews/helpful', { review_id: Number(reviewId) })
      //   .then()
      //   .catch();
    }
  }, [isClicked]);

  const handleClickYes = () => {
    if (!isClicked) {
      setIsClicked(true);
      setCountYes(countYes + 1);
    }
  };

  const handleClickNo = () => {
    if (!isClicked) {
      setIsClicked(true);
      setCountNo(countNo + 1);
    }
  };

  const handleClickReport = () => {
    putReportReview({ review_id: Number(reviewId) }, reviewId);
    reported(reviewId);
  };

  return (
    <div>
      <button
        data-testid="helpfulButton"
        className="helpfulButton"
        type="button"
        onClick={handleClickYes}
      >
        {`Helpful? Yes (${countYes}) `}
      </button>
      <button
        data-testid="nothelpfulButton"
        className="helpfulButton"
        type="button"
        onClick={handleClickNo}
      >
        {` No (${countNo}) `}
      </button>
      <button
        data-testid="report"
        className="helpfulButton"
        type="button"
        onClick={handleClickReport}
      >
        {' | report'}
      </button>
    </div>
  );
};

HelpfulMSG.propTypes = {
  countHelpful: PropTypes.number,
  countNotHelpful: PropTypes.number,
  reviewId: PropTypes.number,
  reported: PropTypes.func,
};

HelpfulMSG.defaultProps = {
  countHelpful: 0,
  countNotHelpful: 0,
  reported: () => 1,
  reviewId: 25711,
};

export default HelpfulMSG;
