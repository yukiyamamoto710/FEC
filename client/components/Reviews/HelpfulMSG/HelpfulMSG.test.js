/**
* @jest-environment jsdom
*/
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  cleanup,
  fireEvent,
  screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import HelpfulMSG from './HelpfulMSG';

afterEach(cleanup);

it('should render corrcet number', () => {
  const test = 23;
  const test1 = 52;
  render(
    <HelpfulMSG
      countHelpful={test}
      countNotHelpful={test1}
    />,
  );
  expect(screen.queryByText('Helpful? Yes (23)')).toBeInTheDocument();
  expect(screen.queryByText('No (52)')).toBeInTheDocument();
  expect(screen.queryByText('| report')).toBeInTheDocument();
});

it('should trigger button to change text one time only', () => {
  const test = 23;
  const test1 = 52;
  const { getByTestId } = render(
    <HelpfulMSG
      countHelpful={test}
      countNotHelpful={test1}
    />,
  );
  fireEvent.click(getByTestId('helpfulButton'));
  fireEvent.click(getByTestId('helpfulButton'));
  expect(getByTestId('helpfulButton')).toHaveTextContent('Helpful? Yes (24)');
});

it('should trigger button to change text one time only', () => {
  const test = 23;
  const test1 = 52;
  const { getByTestId } = render(
    <HelpfulMSG
      countHelpful={test}
      countNotHelpful={test1}
    />,
  );
  fireEvent.click(getByTestId('nothelpfulButton'));
  expect(getByTestId('nothelpfulButton')).toHaveTextContent('No (53)');
  fireEvent.click(getByTestId('nothelpfulButton'));
  expect(getByTestId('nothelpfulButton')).toHaveTextContent('No (53)');
});

it('should trigger button to change text one time only', () => {
  const test = 23;
  const test1 = 52;
  const testfunc = jest.fn();
  const { getByTestId } = render(
    <HelpfulMSG
      countHelpful={test}
      countNotHelpful={test1}
      reported={testfunc}
    />,
  );
  fireEvent.click(getByTestId('report'));
  expect(testfunc).toHaveBeenCalledTimes(1);
});

it('should have default reported', () => {
  expect(HelpfulMSG.defaultProps.reported()).toBeDefined();
});

it('matches snapShot', () => {
  const tree = renderer.create(
    <HelpfulMSG
      countHelpful={23}
      countNotHelpful={52}
      reported={jest.fn()}
      reviewId={25711}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
