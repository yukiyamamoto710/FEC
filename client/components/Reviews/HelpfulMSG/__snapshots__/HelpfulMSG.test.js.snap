// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<div>
  <button
    className="helpfulButton"
    data-testid="helpfulButton"
    onClick={[Function]}
    type="button"
  >
    Helpful? Yes (23) 
  </button>
  <button
    className="helpfulButton"
    data-testid="nothelpfulButton"
    onClick={[Function]}
    type="button"
  >
     No (52) 
  </button>
  <button
    className="helpfulButton"
    data-testid="report"
    onClick={[Function]}
    type="button"
  >
     | report
  </button>
</div>
`;
